source/hwIo/ioThread.py

```diff
@@ -0,0 +1,91 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2016-2022 NV Access Limited, Joseph Lee, Babbage B.V., Davy Kager, Bram Duvigneau,
+# Leonard de Ruijter
+
+import ctypes
+import ctypes.wintypes
+import threading
+import winKernel
+import typing
+from logHandler import log
+import extensionPoints
+import uuid
+
+pre_IoThreadStop = extensionPoints.Action()
+"""
+Executed when the i/o thread is to be stopped.
+This allows components and add-ons to clean up or reset state before background thread shut down.
+Handlers are called with one argument.
+@param ioThread: The thread to shut down
+@type ioThread: IoThread
+"""
+
+
+class IoThread:
+	"""A thread used for background writes and raw I/O, e.g. for braille displays.
+	"""
+
+	thread: typing.Optional[threading.Thread] = None
+	exit: bool = False
+	_apcReferences: typing.Dict[uuid.UUID, winKernel.PAPCFUNC]
+
+	def __init__(self):
+		self._apcReferences = dict()
+
+	def start(self):
+		if self.thread:
+			return
+		thread = self.thread = threading.Thread(
+			name=f"{self.__class__.__module__}.{self.__class__.__qualname__}",
+			target=self.func
+		)
+		thread.daemon = True
+		thread.start()
+		self.handle = ctypes.windll.kernel32.OpenThread(winKernel.THREAD_SET_CONTEXT, False, thread.ident)
+
+	def queueAsApc(
+			self,
+			func: typing.Callable[[int], None],
+			param: int = 0
+	):
+		if not self.handle:
+			raise RuntimeError("IoThread not running")
+
+		# generate an UUID that will be used to cleanup the APC when it is finished
+		apcUuid = uuid.uuid4()
+
+		@winKernel.PAPCFUNC
+		def apc(param):
+			try:
+				if self.exit:
+					return
```

Comment:
can this be moved out of the `try block?

Reply:
Theoretically yes, since the thread is supposed to shut down when exit is True. However, having this in the try block means that the apc is properly removed from the cache, so I'd rather keep it this way.
I initially wrote this without an exception handler actually but decided that it would be better to catch everything, since I'm not sure whether APCs like exceptions very much. Not catching them might be related to the crashes @dkager reported with braille display auto detection. While this is a difference from the original code, the Python thread pool executor also catches exceptions in functions it executes, so I guess it is common sence to do this in our own thread as well.

Reply:
When you say removed from the cache, do you mean ensuring that line `del self._apcReferences[apcUuid]` is called?
If so it might be cleaner to do this with a context manager:
```python
with autoDeleteApcReference(apcUuid):
	if self.exit:
		return
	try:
		func(param)
	except Exception:
		log.error("Error in APC function queued to IoThread", exc_info=True)
```